SimConnect_TransmitClientEvent Implementation

Sector95 opened the pull request:
Example usage, setting the COM1 standby frequency to 127.625 and transponder to 7700:
```rust
use simconnect_sdk::{SimConnect, EventFlag, ClientEventRequest, ClientEvent};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let client = SimConnect::new("Test");

    match client {
        Ok(mut client) => {
            client.subscribe_to_client_event(ClientEventRequest::Com1RadioStbySet)?;
            client.subscribe_to_client_event(ClientEventRequest::TransponderSet)?;

            let com1_event: ClientEvent = ClientEvent::Com1RadioStbySet { value: 123.625 };
            client.transmit_client_event(0, com1_event, EventFlag::DoNothing)?;

            let xpndr_event: ClientEvent = ClientEvent::TransponderSet { value: 7700 };
            client.transmit_client_event(0, xpndr_event, EventFlag::DoNothing)?;
        }
        Err(e) => {
            println!("Error: {e:?}")
        }
    }

    Ok(())
}
```

Sector95 committed:
Implemented TransmitClientEvent call.

Sector95 committed:
Added in transponder set event, updated change log

Sector95 committed:
Added some nav radio event IDs.

mihai-dinculescu commented:
This is very welcome. Thank you!

Unfortunately, the little there is about ClientEvent and NotificationGroup is neither correct nor ergonomic.  Please give me a few days to develop something better to base this PR on.

mihai-dinculescu commented:
OK, it's done. It ended up being a pretty chunky refactor. https://github.com/mihai-dinculescu/simconnect-sdk-rs/commit/d89fa524c2bc585f806f44143a7d95d09040968f.

Probably it's best to read the whole changelog for the rebase, but here are the main takeaways:
- `register_event` has been replaced by `subscribe_to_client_event`. `transmit_client_event` will require the user to call `subscribe_to_client_event` beforehand. It's not ideal, but the SimConnect API is not leaving much choice.
- there is now a `client_event_register` that can be used to determine if the event has been subscribed to, like [here](https://github.com/mihai-dinculescu/simconnect-sdk-rs/blob/main/simconnect-sdk/src/simconnect/events.rs#L105).
- The client event enum is now split into `ClientEventRequest` and `ClientEvent`. Having `transmit_client_event` take a `ClientEvent` will improve the API by making it evident to the user when they have to provide a data field and its purpose.  This [conversion](https://github.com/mihai-dinculescu/simconnect-sdk-rs/blob/main/simconnect-sdk/src/domain/client_event.rs#L171) will probably prove handy to you.

Sector95 commented:
Thanks, I'll take a look and rebase when I get a moment here 👍

SafeShows commented:
Hey @Sector95 How's the feature coming along as I have an app idea that need to write data to sim and or trigger events in the sim

Sector95 commented:
> Hey @Sector95 How's the feature coming along as I have an app idea that need to write data to sim and or trigger events in the sim

Hey sorry, life's been wild, I'll try to tackle this sometime this week!

Sector95 committed:
Merge in event changes.

Sector95 committed:
Cleanup

Sector95 commented:
Updated the example, and merged in your code changes.  Sorry for the delay!

mihai-dinculescu commented on simconnect-sdk/src/simconnect/events.rs:
```diff
@@ -132,4 +132,22 @@ impl SimConnect {
+    pub fn transmit_client_event(
+        &mut self,
+        object_id: u32,
+        event_id: ClientEventRequest,
+        dword: u32,
```
It will be more user-friendly to take `ClientEvent` as a parameter here and use the existing `TryFrom` implementation to split it inside the function into a `ClientEventRequest` and a numeric value.

mihai-dinculescu commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -126,6 +162,40 @@ pub enum ClientEvent {
     /// Toggles parking brake on/off.
     ParkingBrakes,
+    // Comm Radios
+    Com1RadioStbySet {
```
It would be good to add the documentation that's found at https://docs.flightsimulator.com/html/Programming_Tools/Event_IDs/Aircraft_Radio_Navigation_Events.htm#NAV1_STBY_SET_HZ to the events and their inner data.

mihai-dinculescu commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -126,6 +162,40 @@ pub enum ClientEvent {
     ParkingBrakes,
+    // Comm Radios
+    Com1RadioStbySet {
+        value: i32,
```
I don't think that BCD16 values can be negative, so probably this sits better as `u32`.

mihai-dinculescu commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -182,13 +286,39 @@ impl From<ClientEvent> for (ClientEventRequest, i32) {
+    DoNothing,
+    SlowRepeatTimer,
+    FastRepeatTimer,
+    GroupIdIsPriority
```
Probably it's best to skip `GroupIdIsPriority` because it's an advanced use case requiring group priority configuration that's not part of the SDK right now.

mihai-dinculescu commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -182,13 +286,39 @@ impl From<ClientEvent> for (ClientEventRequest, i32) {
+
+/// SimConnect Event Flags for TransmitClientEvent
+/// 
+/// https://docs.flightsimulator.com/html/Programming_Tools/SimConnect/API_Reference/Events_And_Data/SimConnect_TransmitClientEvent.htm
```
This is standard functionality from PREPAR3D, so I guess it makes sense to link to the original documentation https://www.prepar3d.com/SDKv5/sdk/simconnect_api/references/general_functions.html#SimConnect_TransmitClientEvent.

Sector95 replied:
It's a little awkward, as it really dates back to MS FSX.  There's probably going to come a point where Prepar3D and MSFS diverge in the future (if it hasn't already), will be an interesting bridge to cross.

mihai-dinculescu reviewed (commented):
This is coming along nicely. I've left some minor comments.

Sector95 committed:
Changes and BCD16 handling.

Sector95 committed:
conversion logic fix

Sector95 committed:
import cleanup

Sector95 committed:
Changlog fixes

Sector95 commented:
Implemented a little cleaner of an interface for both radios and transponders, so the client doesn't have to do conversion or encoding.  Updated the example on this PR for the new pattern, demonstrating both setting radios and transponders.

Sector95 committed:
loop cleanup and logic fix

Sector95 committed:
formatting fixes.

Sector95 committed:
superficial change for dev clarity

Sector95 commented:
Think I've addressed what showed up in the format workflow there.  Let me know if they're is anything else left outstanding.

mihai-dinculescu reviewed (commented):
Thank you for this.
I've made one final pass and left a few more comments.

mihai-dinculescu commented:
The BCD conversion is excellent! Thanks for adding it!

Sector95 commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -63,6 +85,24 @@ impl ClientEventRequest {
             Self::AxisRightBrakeSet => "AXIS_RIGHT_BRAKE_SET\0".as_ptr() as *const c_char,
             Self::ParkingBrakes => "PARKING_BRAKES\0".as_ptr() as *const c_char,
+            // Comm Radios
+            Self::Com1RadioStbySet => "COM_STBY_RADIO_SET_HZ\0".as_ptr() as *const c_char,
+            Self::Com2RadioStbySet => "COM2_STBY_RADIO_SET_HZ\0".as_ptr() as *const c_char,
+            Self::Com3RadioStbySet => "COM3_STBY_RADIO_SET_HZ\0".as_ptr() as *const c_char,
```
I think if we're going to be opinionated about it, I don't think we should leave open the option for the BCD encoded event personally, even though we have the capability. Hz should be pretty obvious to those trying to set a COM or NAV radio in my opinion. Actually now that I think about it, with our new BCD convertion they'd supply Hz either way.

If you want it to fully spelly "Standby" that's fine, I can make that change.

mihai-dinculescu replied:
It would be good to leave it open for someone to implement it in the future if they want to. I can't see any use cases, but the users of this library always manage to surprise me.
Consistency is more important, however. All the other options in this enum are 1:1 mappings to sim connect and we direct people to read SC's documentation.

Sector95 commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -63,6 +85,24 @@ impl ClientEventRequest {
+            Self::Com1RadioStbySet => "COM_STBY_RADIO_SET_HZ\0".as_ptr() as *const c_char,
+            Self::Com2RadioStbySet => "COM2_STBY_RADIO_SET_HZ\0".as_ptr() as *const c_char,
+            Self::Com3RadioStbySet => "COM3_STBY_RADIO_SET_HZ\0".as_ptr() as *const c_char,
+            Self::Com1RadioSwap => "COM_RADIO_SWAP\0".as_ptr() as *const c_char,
```
Oh, that's interesting... Okay I'll make that change.

Sector95 commented on simconnect-sdk/src/simconnect/events.rs:
```diff
@@ -132,4 +133,29 @@ impl SimConnect {
+    )]
+    pub fn transmit_client_event(
+        &mut self,
+        object_id: u32,
```
Isn't the object_id related to the object you're working with? Ie. 0 is player craft, but above zero might be an AI craft?  Did I misunderstand the docs?

mihai-dinculescu replied:
You're right. It should be internalized and set to 0. The whole library is set up to work just with SIMCONNECT_OBJECT_ID_USER currently. Adding more options is a separate, larger effort.

Sector95 commented on simconnect-sdk/src/domain/client_event.rs:
```diff
@@ -164,31 +258,121 @@ impl TryFrom<&bindings::SIMCONNECT_RECV_EVENT> for ClientEvent {
 }
 
-impl From<ClientEvent> for (ClientEventRequest, i32) {
+impl From<ClientEvent> for (ClientEventRequest, u32) {
```
In this case, the user never actually works with the u32 value, the `ClientEvent` is constructed as such (from the example):

```rust
            let com1_event: ClientEvent = ClientEvent::Com1RadioStbySet { value: 123.625 };
            client.transmit_client_event(0, com1_event, EventFlag::DoNothing)?;
```

The conversion should only be used internally, no?

mihai-dinculescu replied:
> Unless, is this used for when events come back from sim events as well?
Yes, `ClientEventRequest` is used to request events, and `ClientEvent` will be the type of the received events.
So far, I've been sticking with the documented types on the MSFS and SC documentation not to confuse the users.

> The From<ClientEvent> for (ClientEventRequest, u32) should only be used internally, no?
That's the intended use-cause, yes, but it's not possible to enforce [without extra gymnastics](https://users.rust-lang.org/t/visibility-of-trait-implementation/6789/3).

MrMinimal commented:
Given that this isn't merged, does `simconnect-sdk-rs` currently have the ability to transmit client events?

mihai-dinculescu commented:
> Given that this isn't merged, does `simconnect-sdk-rs` currently have the ability to transmit client events and set data (e. g. frequencies) in the sim?

Not without this PR.

@Sector95 are you happy to wrap up the PR? I believe that it's getting very close.
